data/utimer.1
.TH "UTIMER" "1" "June 1, 2009" "" "uTimer"
.SH "NAME"
\fButimer\fP \- command\-line "timer" tool which features a timer, a countdown and a stopwatch.

.SH "SYNOPSIS"
.B utimer
.RI [\-\-timer=TIMELENGTH\ |\ \-t\ TIMELENGTH]
.RI [ option ]...

.B utimer
.RI [\-\-countdown=TIMELENGTH\ |\ \-c\ TIMELENGTH]
.RI [ option ]...

.B utimer
.RI [\-\-stopwatch\ |\ \-s]
.RI [ option ]...

.B utimer
.RI [\-\-limits\ |\ \-L]

.B utimer
.RI [\-\-version]
.SH "DESCRIPTION"
.B utimer
or 
.B µTimer
(pronounced "micro\-timer") is a multifunction timer (command\-line only). It features a timer, a countdown, and a stopwatch.

The
.B timer
counts the elapsed time, increasing to reach the given time length and then stops. Given a time length, the timer will "sleep", wait or pause till the time length has passed, then exits. Compared to the linux
.B sleep
function, µTimer actually shows the current elapsed time which can be useful for long time lengths.

The
.B countdown,
compared to the timer, has the only difference that the starting point is the given time length and it will decrease to reach 0. When it reaches 0 (when the time length has passed), the program exits. It can be very useful for creating scripts.

The
.B stopwatch
is like a real stopwatch where you can measure time of multiple events, or use it for (board) games, or anything else. It starts counting time when the program is started, and it can be stopped (using the 'Q' key) or otherwise you can just hit the ENTER key when you have to take multiple measures without stopping the counting.

.LP 
.SH "OPTIONS"
.B
.IP --help\ |\ \-?
Print the command line options.
.B
.IP --timer=TIMELENGTH\ |\ \-t\ TIMELENGTH
Displays a timer which will run for the specified TIMELENGTH. TIMELENGTH is a pattern that should look like: 2d15h39m58s120ms. See the
.B USAGE
section to read more about how to use the timer.
.B
.IP --countdown=TIMELENGTH\ |\ \-c\ TIMELENGTH
Starts a countdown which will run for the specified TIMELENGTH (from TIMELENGTH down to zero). TIMELENGTH is a pattern that should look like: 2d15h39m58s120ms. See the
.B USAGE
section to read more about how to use the countdown.
.B
.IP --stopwatch\ |\ \-s
Starts the stopwatch and runs until it is stopped by hitting the 'Q' key or by other means (Ctrl+C, system signals, ...). See the
.B USAGE
section to read more about how to use the stopwatch.
.B
.IP --bar
Show a progress bar representing the time left to go. (see also --time and --perc)
.B
.IP --debug\ |\ \-D
Show debug information in output (should ONLY be used for bug reporting, for developers or for testing purposes).
.B
.IP --limits\ |\ \-L
Display the limits of
.B µTimer
for your machine (maximum values, longest timer, ...). It may be not be perfectly accurate. Use this option only to have an idea of what it can do.
.B
.IP --perc
Show a percentage representing the time already elapsed. (see also --time and --bar)
.B
.IP --quiet\ |\ \-q
Quiet or silent output (almost no output) (note that if used with -v, -v is ignored).
.B
.IP --quit-with-success\ |\ \-Q
When you quit the program using the 'q' key, it will exit with a success exit status code, instead of an error one normally. (Note that 'ctrl+c' isn't affected by this option)
.B
.IP --refresh-rate=RATE\ |\ \-r\ RATE
Used to specify a display refresh rate. This purposedly slows down the refresh rate and hides the unneeded time units. RATE can be 'm' for every minute, 's' for every second, 'ms' for every millisecond (actually every ~0.1s). This option can be useful to lower CPU usage or reduce the workload of your terminal (lower erase-and-print frequency). Default is 'ms'.
.B
.IP --time
Show the elapsed/remaining time as text. (Default). See also --bar and --perc.
.B
.IP --no-time
Do not show the elapsed/remaining time (or current time for the stopwatch).
.B
.IP --verbose\ |\ \-v
Print more information during the program's process. (note that if used with -q, this option is ignored)
.SH "USAGE"
.B
.IP Timer:


.B utimer --timer=TIMELENGTH
or
.B utimer -t TIMELENGTH

When using the timer, µTimer will show the current elapsed time and then will exit when TIMELENGTH is reached (read: when that TIMELENGTH has passed).
The TIMELENGTH is composed of pairs of: a value + a suffix (for example "39d" for 39 days). It can be in any order, and there is no space between each pair. Each value 
.B must be a positive integer
and can be null (0). The suffix is case
.B sensitive
(see below for the suffixes list). 

Here's a few valid examples of TIMELENGTH:
55s,
2m450ms,
0s,
0d32h,
9m7h30s80ms6d,
1d0h0m50s.

Supported suffixes are:
.B d
for days,
.B h
for hours,
.B m
for minutes,
.B s
for seconds,
.B ms
for milliseconds.

.B
.IP Countdown:

.B utimer --countdown=TIMELENGTH
or
.B utimer -c TIMELENGTH

When using the countdown, µTimer will show the current remaining time before exiting. After TIMELENGTH has passed -or when the remaining time reaches zero (0)- µTimer exits.
See the Timer usage to know how to write a valid TIMELENGTH.
.B
.IP Stopwatch:

.B utimer --stopwatch
or
.B utimer -s

When using the stopwatch, it counts until it is stopped using the 'Q' key or Ctrl+C or certain system signals (SIGKILL for example).
You can use the ENTER key to take measures (like a snapshot) of the elapsed time.


.SH SHORTCUT KEYS

.B
.IP 'Q':


You can stop/quit the timer whenever by using the 'Q' key, note that the program will exit with a error exit status code (1) (you can bypass this behaviour by using the
.B --quit-with-success
option, see the
.B EXIT_STATUS
section above).

.B
.IP SPACEBAR:

You can pause the timer anytime using the spacebar key. To resume, just hit the spacebar again.

.B
.IP ENTER:

With the stopwatch, you can use the ENTER key to take measures (like a snapshot) of the elapsed time.

.SH EXIT STATUS
After the timer or the countdown are done, the program exits with a success exit status code (0). But when you quit during the timer/countdown is counting using 'q', it exits with an error exit status code (1). You can change this behaviour using the 
.B --quit-with-success
option. (Note that 'ctrl+c' isn't affected by this option)

.SH "EXAMPLE"
.IP Timer:


.B utimer \-t 20s900ms
runs the timer for 20 seconds and 900 milliseconds.
.IP
.B utimer \--timer=1d
runs the timer for a whole day.

.IP Countdown:

.B utimer \-c 1m48s500ms
runs the countdown for 1 minute, 48 seconds and 500 milliseconds.
.IP
.B utimer \--countdown=2d5m
runs the countdown for 2 days plus 5 minutes.

.IP Stopwatch:

.B utimer \-s
starts the stopwatch.


.SH "ACCURACY"
The timer can be accurate enough for most common uses, and we cannot give any warranty on its accuracy. This is due to the time needed to start and to exit the program. On our test machines, it's as low as 5 milliseconds.
.SH "BUGS"
If you find any bug using µTimer, please report it to:

.B https://bugs.launchpad.net/utimer

.B or
send it to:

.B bugs@utimer.codealpha.net
.SH "AUTHOR"
Arnaud "Weboide" Soyez <weboide@codealpha.net>


src/utimer.c
/*
 *  utimer.c
 *
 *  Copyright 2008-2010  Arnaud Soyez <weboide@codealpha.net>
 *
 *  This file is part of uTimer.
 *  (uTimer is a CLI program that features a timer, countdown, and a stopwatch)
 *
 *  uTimer is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  uTimer is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with uTimer.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */

#include "utimer.h"

static gchar *timer_info, *countdown_info, *refresh_rate;
static gboolean stopwatch = FALSE,
        show_limits = FALSE,
        show_version = FALSE,
        show_bar = FALSE,
        show_perc = FALSE,
        show_text = TRUE;

static GOptionEntry entries[] = {

  {"bar",
   0,
   0,
   G_OPTION_ARG_NONE,
   &show_bar,
   N_("show a progress bar representing the remaining/elapsed time"),
   NULL},

  {"countdown",
   'c',
   0,
   G_OPTION_ARG_STRING,
   &(countdown_info),
   N_("count from TIMELENGTH down to 0 then exit (e.g. -c 30d9h50s)"),
   N_("TIMELENGTH")},

  {"debug",
   'd',
   0,
   G_OPTION_ARG_NONE,
   &(ut_config.debug),
   N_("debug output"),
   NULL},

  {"limits",
   'L',
   0,
   G_OPTION_ARG_NONE,
   &(show_limits),
   N_("show the limits of µTimer"),
   NULL},

  {"perc",
   0,
   0,
   G_OPTION_ARG_NONE,
   &show_perc,
   N_("show a percentage representing the elapsed time"),
   NULL},

  {"quiet",
   'q',
   0,
   G_OPTION_ARG_NONE,
   &(ut_config.quiet),
   N_("quiet output"),
   NULL},

  {"quit-with-success",
   'Q',
   0,
   G_OPTION_ARG_NONE,
   &(ut_config.quit_with_success),
   N_("return a success exit status (0) when using 'q' to exit"),
   NULL},

  {"refresh-rate",
   'r',
   0,
   G_OPTION_ARG_STRING,
   &(refresh_rate),
   N_("display refresh rate. RATE can be 'm' for minute, 's' for\
 second, 'ms' for millisecond (default)"),
   N_("RATE")},

  {"stopwatch",
   's',
   0,
   G_OPTION_ARG_NONE,
   &(stopwatch),
   N_("start the stopwatch. Hit 'Q' to exit"),
   NULL},

  {"time",
   0,
   0,
   G_OPTION_ARG_NONE,
   &show_text,
   N_("show elapsed/remaining time as text (default)"),
   NULL},

  {"no-time",
   0,
   G_OPTION_FLAG_REVERSE,
   G_OPTION_ARG_NONE,
   &show_text,
   N_("do not show the text telling the elapsed/remaining time (see --time)"),
   NULL},

  {"timer",
   't',
   0,
   G_OPTION_ARG_STRING,
   &(timer_info),
   N_("count from 0 to TIMELENGTH then exit (e.g. -t 31m27s300ms)"),
   N_("TIMELENGTH")},

  {"verbose",
   'v',
   0,
   G_OPTION_ARG_NONE,
   &(ut_config.verbose),
   N_("verbose output"),
   NULL},

  {"version",
   'V',
   0,
   G_OPTION_ARG_NONE,
   &(show_version),
   N_("display the current version of µTimer"),
   NULL},

  {NULL}
};

/**
 * Activate/Deactivate the canonical mode from a TTY.
 */
void set_tty_canonical(int state)
{
  struct termios ttystate;

  if (state == 1)
  {
    g_debug("Activating canonical mode.");
    tcgetattr(STDIN_FILENO, &ttystate);
    ttystate.c_lflag &= ~ICANON; // remove canonical mode
    ttystate.c_cc[VMIN] = 1; // minimum length to read before sending
    tcsetattr(STDIN_FILENO, TCSANOW, &ttystate); // apply the changes
  }
  else
  {
    g_debug("Deactivating canonical mode.");
    tcsetattr(STDIN_FILENO, TCSANOW, &savedttystate); // put canonical mode back
  }
}

/**
 * Deactivate the canoncial mode (used with atexit())
 */
void reset_tty_canonical_mode()
{
  set_tty_canonical(0);
}

/**
 * Quits the main loop to end the program.
 * Quits the main loop to end the program with error_status
 * (EXIT_SUCCESS or EXIT_FAILURE).
 */
void quitloop(int error_status)
{
  g_assert(loop);
  g_debug("%s: Stopping Main Loop (error code: %i)...", __FUNCTION__, error_status);

  // We set the exit status code
  ut_config.current_exit_status_code = error_status;
  g_main_loop_quit(loop);
}

/**
 * Quits the main loop with error to end the program.
 * Quits the main loop to end the program ungracefully (EXIT_FAILURE).
 * This function just calls quitloop(EXIT_FAILURE).
 * This is needed for signal traps.
 */
void error_quitloop()
{
  quitloop(EXIT_FAILURE);
}

/**
 * Quits the main loop with success code to end the program.
 * Quits the main loop to end the program gracefully (EXIT_SUCCESS).
 * This function just calls quitloop(EXIT_SUCCESS).
 * This is needed for callbacks.
 */
void success_quitloop()
{
  quitloop(EXIT_SUCCESS);
}

/**
 * Updates terminal_cols and starts waiting for signal again.
 */
void terminal_size_changed(gint s)
{
  ut_config.terminal_cols = get_terminal_width();
  g_message("\r ");
  g_debug("%s: Received SIGWINCH cols=%d",
          __FUNCTION__, ut_config.terminal_cols);
}

/**
 * Check to see if the user wants to quit.
 * This function waits till the user hits the 'q' key to quit the program,
 * then it will call the quitloop function. This is called by a thread
 * to avoid blocking the program.
 */
int check_exit_from_user()
{
  set_tty_canonical(1); /* Apply canonical mode to TTY*/
  g_atexit(reset_tty_canonical_mode); /* Deactivate canonical mode at exit */

  gint c;
  do
  {
    c = fgetc(stdin);
    g_print("\b \b"); // backspace, write a space to clear, backspace again
    switch (c)
    {
      case ' ':
      {
        if (paused)
        {
          g_timer_continue(ut_config.timer);
          paused = FALSE;
        }
        else
        {
          g_timer_stop(ut_config.timer);
          paused = TRUE;
        }
      }

    }
  } while (c != 'q' && c != 'Q'); /* checks for 'q' key */

  /* If the user asks for exiting, we stop the loop. */
  quitloop((ut_config.quit_with_success ? EXIT_SUCCESS : EXIT_FAILURE));
}

static void clean_up(void)
{
  free_config(&ut_config);

  if (countdown_info)
  {
    g_debug("Freeing countdown_info...");
    g_free(countdown_info);
    countdown_info = NULL;
  }

  if (timer_info)
  {
    g_debug("Freeing timer_info...");
    g_free(timer_info);
    timer_info = NULL;
  }

  if (refresh_rate)
  {
    g_debug("Freeing refresh_rate...");
    g_free(refresh_rate);
    refresh_rate = NULL;
  }
}

int main(int argc, char *argv[])
{ /* ================== MAIN STARTS ==================== */
  GError *error = NULL;
  GOptionContext *context;
  gchar *tmp = NULL;
  ut_timer *ttimer;
  gint print_refresh_rate;
  /* -------------- Initialization ------------- */

  tcgetattr(STDIN_FILENO, &savedttystate); /* Save current tty state  */

  g_thread_init(NULL);
  g_type_init();
  init_config(&ut_config);

  /* We cleanup at exit */
  g_atexit(clean_up);

  /* Define localization */
  ut_config.locale = setlocale(LC_ALL, "");
  if (!ut_config.locale)
  {
    g_error(_("Error during setting current locale. Falling back to default locale."));
    ut_config.locale = setlocale(LC_ALL, "C");
    if (!ut_config.locale)
    {
      g_critical(_("Couldn't set any locale. Exiting..."));
      exit(EXIT_FAILURE);
    }
  }

  /* i18n */
  g_set_prgname(PACKAGE);
  bindtextdomain(GETTEXT_PACKAGE, LOCALEDIR);
  bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
  textdomain(GETTEXT_PACKAGE);

  /* Set up the log handler */
  setup_log_handler();

  /* Handles any change of size for the terminal */
  (void) signal(SIGWINCH, terminal_size_changed);
  /* force the signal to get the current terminal size */
  terminal_size_changed(0);

  /* Set the function to call in case of receiving signals:
   * we need to stop the main loop to exit correctly, but
   * with error code!
   */
  (void) signal(SIGALRM, error_quitloop);
  (void) signal(SIGHUP, error_quitloop);
  (void) signal(SIGINT, error_quitloop);
  (void) signal(SIGPIPE, error_quitloop);
  (void) signal(SIGQUIT, error_quitloop);
  (void) signal(SIGTERM, error_quitloop);



  /* ------------ Initialization is done ---------- */

  /* -------------- Options parsing ------------- */

  tmp = g_strconcat(_("[ARGUMENTS] - "), SHORTDESCRIPTION, NULL);
  context = g_option_context_new(tmp);
  g_free(tmp);

  g_option_context_set_summary(context, SUMMARY);

  tmp = g_strconcat(DESCRIPTION,
                    _("\nReport any bug to: https://bugs.launchpad.net/utimer/+filebug"),
                    NULL);
  g_option_context_set_description(context, tmp);
  g_free(tmp);

  g_option_context_add_main_entries(context, entries, NULL);

  if (!g_option_context_parse(context, &argc, &argv, &error))
  {
    g_printerr(_("Error while parsing options: %s.\n"), error->message);
    g_printerr(_("Run '%s --help' to see a full list of available command\
 line options.\n"), argv[0]);
    g_error_free(error);
    exit(EXIT_FAILURE);
  }

  g_option_context_free(context);


  /* ------------ Options parsing DONE ------------- */


  /* Verify that there is at least an option */
  if (!(timer_info
        || show_version
        || show_limits
        || countdown_info
        || stopwatch))
  {
    g_printerr(_("No main option (-t, -c or -s) has been specified!\n"));
    g_printerr(_("Run '%s --help' to see a full list of available command\
 line options.\n"), argv[0]);
    exit(EXIT_FAILURE);
  }


  /* Check if there is any conflicting options */

  if (ut_config.debug)
    ut_config.quiet = FALSE;

  if (timer_info && (countdown_info || stopwatch)
      || countdown_info && stopwatch)
  {
    g_warning(_("Conflicting options!\nThe following options cannot\n\
 be used simultaneously:\n -t (timer mode), -c (countdown mode), -s\
 (stopwatch mode)."));
    exit(EXIT_FAILURE);
  }

  /* set up refresh rate if given */
  print_refresh_rate = 89;
  if (refresh_rate)
  {
    if (g_ascii_strcasecmp(refresh_rate, "m") == 0)
    {
      print_refresh_rate = 60 * 1000;
    }
    else if (g_ascii_strcasecmp(refresh_rate, "s") == 0)
    {
      print_refresh_rate = 1000;
    }
  }

  g_debug("G_MAXUINT = %u", G_MAXUINT);

  if (isatty(STDOUT_FILENO))
    g_debug("\033[34mYou are using a TTY.\033[m");
  else
    g_debug("You are not using a TTY.");

  if (ut_config.debug)
  {
    g_debug("Deactivating buffered output for stderr and stdout...");

    setbuf(stdout, NULL);
    setbuf(stderr, NULL);
  }

  /* Showing limits */

  if (show_limits)
  {
    g_print(_("This is a list of the possible limits of µTimer for your\
 machine.\n\n"));

    tmp = timer_get_maximum_time();
    g_print(_("* Limits for the Timer/Countdown:\n"));
    g_print(_("\t- The maximum possible timer length is: %s.\n"), tmp);
    g_print(_("\t  If you enter a value that is exceeding it, it will be\
 replaced by the value above.\n"));

    g_free(tmp);
    tmp = NULL;

    exit(EXIT_SUCCESS);
  }

  /* Showing version */

  if (show_version)
  {
    g_print("%s\n\n%s\n\n%s\n", PACKAGE_STRING, COPYRIGHT, COPYRIGHT_TXT);
    exit(EXIT_SUCCESS);
  }

  /* ------------------ PROCESSING STARTS ------------------ */

  /* Prepare for starting the main loop */

  loop = g_main_loop_new(NULL, FALSE);

  /* -------------- TIMER & COUNTDOWN MODE -------------- */
  if (timer_info || countdown_info || stopwatch)
  {
    g_debug("Setting up default precision");
    timer_precision precision = TIMER_PRECISION_MILLISECOND;
    g_debug("Setting up default display");
    g_debug("perc %i", show_perc);
    g_debug("text %i", show_text);
    g_debug("bar %i", show_bar);
    timer_display options_timer_display = { .perc = show_perc,
                                            .text = show_text,
                                            .bar = show_bar };

    /* set up refresh rate if given */
    if (refresh_rate)
    {
      if (g_ascii_strcasecmp(refresh_rate, "m") == 0)
      {
        precision = TIMER_PRECISION_MINUTE;
      }
      else if (g_ascii_strcasecmp(refresh_rate, "s") == 0)
      {
        precision = TIMER_PRECISION_SECOND;
      }
    }

    guint seconds = 0;
    guint mseconds = 0;

    if (countdown_info)
    {
      g_debug("Countdown Mode");
      parse_time_pattern(countdown_info, &(seconds), &(mseconds));
      ttimer = timer_new_countdown(seconds, mseconds, success_quitloop, error_quitloop, ut_config.timer, precision, &options_timer_display);
    }
    else if (timer_info)
    {
      g_debug("Timer Mode");
      parse_time_pattern(timer_info, &(seconds), &(mseconds));
      ttimer = timer_new_timer(seconds, mseconds, success_quitloop, error_quitloop, ut_config.timer, precision, &options_timer_display);
    }
    else
    {
      g_debug("Stopwatch Mode");
      ttimer = timer_new_stopwatch(success_quitloop, error_quitloop, ut_config.timer, precision, &options_timer_display);
    }

    tmp = timer_get_maximum_time();
    g_debug("Maximum Timer length is %s.", tmp);
    g_free(tmp);
    tmp = NULL;

    if (!ut_config.quiet && ut_config.verbose && !stopwatch)
    {
      tmp = timer_ut_timer_to_string(ttimer);
      g_info(_("Timer will stop in approximately %s. Hit 'Q' to abort."), tmp);
      g_free(tmp);
      tmp = NULL;
    }

    // print timer once because the timeout loop waits before executing the func
    timer_print(ttimer);
    // create the timeout loop to print the timer every refresh_rate
    ttimer->timer_print_source_id = g_timeout_add(print_refresh_rate,
                                                  (GSourceFunc) timer_print,
                                                  ttimer);
    if (ttimer->mode != TIMER_MODE_STOPWATCH)
    {
      g_debug("Starting Timer thread");

      if (!g_thread_create((GThreadFunc) timer_check_loop, ttimer, FALSE, &error))
      {
        g_printerr(_("Thread creation failed: %s"), error->message);
        g_error_free(error);
        error_quitloop();
        exit(EXIT_FAILURE);
      }
    }
  } /* -------------- END TIMER & COUNTDOWN MODE -------------- */
  else
  {
    /* No mode selected! We quit the loop ASAP. */
    g_idle_add((GSourceFunc) error_quitloop, NULL);
  }

  g_debug("Creating thread exit check");
  if (!g_thread_create((GThreadFunc) check_exit_from_user, NULL, FALSE, &error))
  {
    // thread creation failed!
    g_printerr(_("Thread creation failed: %s"), error->message);
    g_error_free(error);
    exit(EXIT_FAILURE);
  }

  /* ------------- MAIN LOOP ---------------- */

  g_debug("Starting main loop...");
  g_main_loop_run(loop);
  g_main_loop_unref(loop);
  loop = NULL;
  g_debug("Exiting main loop...");

  /* Print the timer one more time to show the actual time (in case of slow
   * refresh rates. */
  timer_print(ttimer);


  /* ------------- END OF MAIN LOOP ---------------- */

  g_debug("Quitting with error code: %i", ut_config.current_exit_status_code);


  /* ================== CLEAN UP ==================== */
  g_message("\n"); // print a new line if not in quiet mode
  timer_destroy(ttimer);

  /* ================== MAIN DONE ==================== */
  return ut_config.current_exit_status_code;
}
